qa/python_models/python_based_backends/add_sub_backend/model.py

```diff
@@ -0,0 +1,121 @@
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import json
+import os
+
+import triton_python_backend_utils as pb_utils
+
+_ADD_SUB_ARGS_FILENAME = "model.json"
+
+
+class TritonPythonModel:
+    def initialize(self, args):
+        """This function allows the model to initialize any state associated with this model.
+
+        Parameters
+        ----------
+        args : dict
+          Both keys and values are strings. The dictionary keys and values are:
+          * model_config: A JSON string containing the model configuration
+          * model_instance_kind: A string containing model instance kind
+          * model_instance_device_id: A string containing model instance device ID
+          * model_repository: Model repository path
+          * model_version: Model version
+          * model_name: Model name
+        """
+
+        self.model_config = model_config = json.loads(args["model_config"])
+
+        # Get OUTPUT configuration
+        output_config = pb_utils.get_output_config_by_name(model_config, "OUTPUT")
+
+        engine_args_filepath = os.path.join(
+            pb_utils.get_model_dir(), _ADD_SUB_ARGS_FILENAME
+        )
+        assert os.path.isfile(
+            engine_args_filepath
+        ), f"'{_ADD_SUB_ARGS_FILENAME}' containing add sub model args must be provided in '{pb_utils.get_model_dir()}'"
+
+        with open(engine_args_filepath) as file:
+            self.add_sub_config = json.load(file)
+
+        assert (
+            "operation" in self.add_sub_config
+        ), f"Missing required key 'operation' in {_ADD_SUB_ARGS_FILENAME}"
+
+        extra_keys = set(self.add_sub_config.keys()) - {"operation"}
+        assert (
+            not extra_keys
+        ), f"Unsupported keys are provided in {_ADD_SUB_ARGS_FILENAME}: {', '.join(extra_keys)}"
+
+        assert self.add_sub_config["operation"] in [
+            "add",
+            "sub",
+        ], f"'operation' value must be 'add' or 'sub' in {_ADD_SUB_ARGS_FILENAME}"
+
+        # Convert Triton types to numpy types
+        self.output_dtype = pb_utils.triton_string_to_numpy(output_config["data_type"])
+
+    def execute(self, requests):
+        """This function is called when an inference request is made
+        for this model.
+
+        Parameters
+        ----------
+        requests : list
+          A list of pb_utils.InferenceRequest
+
+        Returns
+        -------
+        list
+          A list of pb_utils.InferenceResponse. The length of this list must
+          be the same as `requests`
+        """
+
+        responses = []
+
+        for request in requests:
+            in_0 = pb_utils.get_input_tensor_by_name(request, "INPUT0")
+            in_1 = pb_utils.get_input_tensor_by_name(request, "INPUT1")
+
+            if self.add_sub_config["operation"] == "add":
```

Comment:
I believe we can eliminate dependency on single output [here](https://github.com/triton-inference-server/server/pull/6466/files#diff-c02333b955665a5e56699cccc5cce927d0c8ec9ac1fe6a783dc2fd3cc8a3e802R57), if we make this model to output both add tensor and sub tenor. In this case we can also shorten our `initialize` function to get rid of some checks and `add_model.json` and `sub_model.json` . CC @tanmayv25  what do you think?

Reply:
I would also suggest adding second version to at least one model. It can be the same e.g. pytorch model, just in a separate version folder

Reply:
The reason for "operation" field in the add_sub json model is to test whether two models belonging to the same python based backend can be loaded and served. In order to clean up the dependency of single output, we can do the following:

1. variant1.json model: returns sum in OUTPUT0 and difference in OUTPUT1.
2. variant2.json model: returns difference in OUTPUT0 and sum in OUTPUT1.

What are your thoughts?

Reply:
I am fine with keeping this as is, if this is too much work. I thought we can also add some versions to some models

Reply:
@oandreeva-nv @tanmayv25,
Thank you. The above changes may still require some conditional checks in the model.py. Therefore, proceeding with the existing approach. Updated the add_sub backend model.py to prepare configuration using the auto-config feature and removed a separate config.pbtxt file.
In addition, removed add_model.json, and sub_model.json files and creating them using a simple echo command in the test.sh file.
```
echo '{ "operation": "add" }' > ${MODEL_REPOSITORY}/add/1/model.json
```

Also, added multiple versions for the add_sub backend (add) and python backend models, along with appropriate tests to ensure expected behavior.